### PR TITLE
Refs #28459 -- Improved performance of BaseExpression.convert_value().

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://code.djangoproject.com/ticket/28459

Before:
```
In [2]: %timeit for x in City.objects.annotate(v=models.Value(1, output_field=models.IntegerField())).values_list('v'): pass
262 ms ± 3.42 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
In [3]: %timeit for x in City.objects.annotate(v=models.Value('1', output_field=models.TextField())).values_list('v'): pass
243 ms ± 3.77 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
After:
```
In [2]: %timeit for x in City.objects.annotate(v=models.Value(1, output_field=models.IntegerField())).values_list('v'): pass
191 ms ± 2.08 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
In [3]: %timeit for x in City.objects.annotate(v=models.Value('1', output_field=models.TextField())).values_list('v'): pass
65.6 ms ± 878 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

### CODE REVIEW

**Comment by timgraham** [comment]
`Expression.convert_value()` is documented so I don't think this change is acceptable:
`ref/models/expressions.txt:    .. method:: convert_value(value, expression, connection)`

**Comment by sir-sigurd** [comment]
I think I will propose to deprecate `convert_value()` in the future but for now I'd like to not break it.
If you think that the current approach is breaking I could rework it in compatible way.

**Comment by timgraham** [comment]
The current approach changes `convert_value()` from a method accepting arguments to a cached property, so Expression subclasses overriding it will break, correct?

**Comment by sir-sigurd** [comment]
But this property returns callable. For example:
```
In [6]: class A:
   ...:     @cached_property
   ...:     def method(self):
   ...:         return lambda arg: 'property ' + arg
   ...:
In [11]: class B(A):
    ...:     def method(self, arg):
    ...:         return 'method ' + arg + ' ' + super().method(arg)
    ...:
In [12]: B().method('test')
Out[12]: 'method test property test'
```

I see the only difference in behavior: currently `Expression.convert_value()` is a function that could be called if supply instance to along with other arguments.

**Comment by timgraham** [comment]
I see now (and of course there are expressions in Django that show backwards compatibility is okay). Well, perhaps it would be good to add some comments explaining why things are implemented the way they are because IMO this obscures the implementation quite a bit. If we proceed with the patch, it might be useful to avoid the repetition of `lambda value, expression, connection: None if value is None` if possible. \cc @jarshwah for input.

File: django/db/models/expressions.py [hunk 1]
```diff
@@ -274,23 +284,26 @@ def _resolve_output_field(self):
...
         from the one the database returns.
         """
         field = self.output_field
         internal_type = field.get_internal_type()
-        if value is None:
-            return value
-        elif internal_type == 'FloatField':
-            return float(value)
+        if internal_type == 'FloatField':
+            return lambda value, expression, connection: None if value is None else float(value)
```
Comment (millerdev) [inline django/db/models/expressions.py]: @sir-sigurd these lambdas break expression pickling:
```
Traceback (most recent call last):
  File "/home/travis/build/millerdev/django-cte/tests/test_recursive.py", line 212, in test_pickle_recursive_cte_queryset
    pickled_qs = pickle.loads(pickle.dumps(regions))
AttributeError: Can't pickle local object 'BaseExpression.convert_value.<locals>.<lambda>'
```
Found here: https://travis-ci.org/millerdev/django-cte/jobs/366128465

Comment (timgraham) [reply django/db/models/expressions.py]: Please report bugs at https://code.djangoproject.com/.

Comment (millerdev) [reply django/db/models/expressions.py]: False alarm. Looks like the [custom `__getstate__`](https://github.com/django/django/pull/8854/files#diff-fe637a3ec6af9cbe2ee87293000e7c82R150) has subsequently been removed, which is the likely cause of the issue. Tracking that down now.

**Pull request merged**